### PR TITLE
MOTS-242: Unassigning module should not be possible when module is in progress

### PR DESCRIPTION
**Pull request by dserkowski**

### CODE REVIEW

File: src/main/java/org/motechproject/mots/repository/ModuleRepository.java [hunk 1]
```diff
@@ -1,11 +1,19 @@
...
 import org.motechproject.mots.domain.Module;
+import org.springframework.data.jpa.repository.Query;
 import org.springframework.data.repository.CrudRepository;
+import org.springframework.data.repository.query.Param;
 
 public interface ModuleRepository extends CrudRepository<Module, UUID> {
 
   Optional<Module> findById(UUID id);
+
+  @Query("SELECT DISTINCT mp.module FROM ModuleProgress mp "
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/repository/ModuleRepository.java]: you don't need a querry here, if you rename the method to findByCommunityHealthWorkerIdAndStatusNot and add status parameter.
you can remove this method, see the comments below

File: src/main/java/org/motechproject/mots/web/ModuleAssignmentController.java [hunk 2]
```diff
@@ -67,4 +67,19 @@ public ChwModulesDto getAssignedModules(@RequestParam(value = "chwId") String ch
...
+
+  /**
+   * Get Modules with started CHW ModuleProgress assigned to CHW.
+   * @param chwId id of CHW
+   * @return modules assigned to CHW
+   */
+  @RequestMapping(value = "/startedAssignedModules", method = RequestMethod.GET)
+  @ResponseStatus(HttpStatus.OK)
+  @ResponseBody
+  public ChwModulesDto getStartedAssignedModules(@RequestParam(value = "chwId") String chwId) {
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/web/ModuleAssignmentController.java]: remove this, later to recognize the not started modules on the UI, the ChwModulesDto will be changed to store this information

File: src/main/java/org/motechproject/mots/service/ModuleAssignmentService.java [hunk 3]
```diff
@@ -199,6 +229,19 @@ public void assignModulesToDistrict(DistrictAssignmentDto assignmentDto) {
     }
   }
 
+  /**
+   * Get AssignedModules consists of Modules with started CHW ModuleProgress.
+   * @param chwId Id of CHW
+   * @return modules assigned to CHW with given Id
+   */
+  @PreAuthorize(RoleNames.HAS_ASSIGN_MODULES_ROLE)
+  public AssignedModules getStartedAssignedModules(UUID chwId) {
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/service/ModuleAssignmentService.java]: remove this, see the comment below

File: src/main/java/org/motechproject/mots/service/ModuleAssignmentService.java [hunk 4]
```diff
@@ -97,9 +107,10 @@ public void assignModules(AssignedModules assignedModules) {
 
     Set<Module> modulesToAdd = getModulesToAdd(oldModules, newModules);
     Set<Module> modulesToRemove = getModulesToRemove(oldModules, newModules);
+    validateModulesToUnassign(assignedModulesChw, modulesToRemove);
+    moduleProgressService.removeModuleProgresses(assignedModulesChw, modulesToRemove);
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/service/ModuleAssignmentService.java]: move removeModuleProgresses at the end of this method next to createModuleProgresses

File: src/main/java/org/motechproject/mots/service/ModuleAssignmentService.java [hunk 5]
```diff
@@ -114,7 +125,26 @@ public void assignModules(AssignedModules assignedModules) {
       throw new ModuleAssignmentException(message, ex);
     }
 
-    moduleProgressService.createModuleProgresses(assignedModules.getHealthWorker(), modulesToAdd);
+    moduleProgressService.createModuleProgresses(assignedModulesChw, modulesToAdd);
+  }
+
+  private void validateModulesToUnassign(CommunityHealthWorker chw, Set<Module> modulesToUnassing) {
+    HashSet<Module> startedModules = new HashSet<>();
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/service/ModuleAssignmentService.java]: change it to `Set<String> startedModulesIds`

File: src/main/java/org/motechproject/mots/service/ModuleAssignmentService.java [hunk 6]
```diff
@@ -114,7 +125,26 @@ public void assignModules(AssignedModules assignedModules) {
...
+        .ifPresent(moduleProgress -> {
+          if (!ProgressStatus.NOT_STARTED.equals(moduleProgress.getStatus())) {
+            startedModules.add(module);
+          }
+        })
+    );
+
+    if (!startedModules.isEmpty()) {
+      throw new MotsException(String.format("Could not unassign started modules (module IDs: %s)",
+          startedModules.stream()
```
Comment (pmuchowski) [inline src/main/java/org/motechproject/mots/service/ModuleAssignmentService.java]: use StringUtils.join here

File: README.md [hunk 7]
```diff
@@ -32,10 +32,9 @@ Under File -> Settings -> Editor -> Code Style import the google-styleguide (gea
 ## Loading initial test data
 1. In `mots` root directory run:
 ```
-mysql -uroot -p mots < src/main/resources/test_modules.sql
-mysql -uroot -p mots < src/main/resources/test_ivr_config.sql
+mysql -uroot -ppassword mots < src/main/resources/test_modules.sql
+mysql -uroot -ppassword mots < src/main/resources/test_ivr_config.sql
```
Comment (kkaczmarczyk) [inline README.md]: let's remove it - password can be different on each env

File: src/main/java/org/motechproject/mots/repository/ModuleRepository.java [hunk 8]
```diff
@@ -1,11 +1,19 @@
...
+      + "WHERE mp.communityHealthWorker.id = :chwId "
+      + "AND mp.status <> org.motechproject.mots.domain.enums.ProgressStatus.NOT_STARTED ")
+  Set<Module> findByCommunityHealthWorkerIdAndStatus(@Param("chwId") UUID chwId);
```
Comment (kkaczmarczyk) [inline src/main/java/org/motechproject/mots/repository/ModuleRepository.java]: Can this be done using regular hibernate methods? Have a look at CommunityHealthWorkerRepository.findByCommunityFacilityChiefdomDistrictId

File: README.md (hunk 7 above)
Comment (dserkowski) [reply README.md]: Ok

**Pull request merged**